[Gu Geng] iP

jackgugz opened the pull request:
# DUKE the Super-Duper

> Free thing is great thing. 
> -[Jack](https://github.com/jackgugz) 

### Duke frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use
<img src="https://www.vippng.com/png/detail/7-77855_harold-thumbs-up-harold-thumbs-up-transparent.png" alt="Thumb up image" width="200"/>

### All you need to do is,

1. download it from [here](https://github.com/jackgugz/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 👍 
And it is **FREE**! ⚡ 

Features:

 - [X] Managing todos
 - [X] Managing deadlines
 - [X] Managing events
 - [X] Find tasks on specified date
 - [X] Find tasks with specified keywords

#
If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

jackgugz committed:
echo added

jackgugz committed:
add ability to store the input and display when requested

jackgugz committed:
add ability to mark tasks as done

jackgugz committed:
add support for tracking todos, deadlines and events

jackgugz committed:
automated text UI testing

jackgugz committed:
teach duke to deal with errors

jackgugz committed:
Add support for deleting tasks from the list.

jackgugz committed:
used enums and updated the testing cases

jackgugz committed:
formatting cleaned up

jackgugz committed:
Add save function to Duke

jackgugz committed:
Add date recognition and date filter function

jackgugz committed:
Merge branch 'branch-level-7'

jackgugz committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Duke.java
#	src/main/java/Event.java
#	src/main/java/Task.java
#	src/main/java/Todo.java

jackgugz committed:
Refactor the code to extract out closely related code as classes.

jackgugz committed:
Divide classes into packages

jackgugz committed:
Add Junit tests
Classes with Junit test:
duke.Duke
duke.task.Deadline
duke.command.AddCommand

jackgugz committed:
Package the App as a JAR file

jackgugz committed:
Add JavaDoc comments

jackgugz committed:
Tweak the code to comply with a coding standard

jackgugz committed:
Give users a way to find a task by searching for a keywor

jackgugz committed:
Merge branch 'branch-A-JavaDoc'

jackgugz committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/TaskList.java

jackgugz committed:
Merge branch 'branch-Level-9'

jackgugz committed:
Adjust JavaDoc and Coding Standard for newly added methods/class
duke.command.FindCommand
duke.Ui::showFindList
duke.TaskList::listFind

jackgugz committed:
Fix bug of rendering all tasks with time without proper filtering

jackgugz committed:
Merge remote-tracking branch 'origin/add-gradle-support'

jackgugz committed:
Automate project builds using Gradle

jackgugz committed:
Merge branch 'branch-A-Gradle'

jackgugz committed:
Use checkStyle to detect coding style violations

jackgugz committed:
Merge branch 'branch-A-CheckStyle'

jackgugz committed:
Earn JavaFX basics

jackgugz committed:
Create a GUi for Duke

jackgugz committed:
Merge branch 'branch-Level-10'

jackgugz committed:
Use Java varargs feature

jackgugz committed:
Merge branch 'branch-A-Varargs'

jackgugz committed:
Update test classes due to changes to the Task classes

Jeromesyl commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,116 @@
+            fw.close();
+            Files.delete(filePath);
+            Files.copy(temp, filePath);
+            //noinspection ResultOfMethodCallIgnored
```
What does this comment mean? Should it be included?

Jeromesyl commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,140 @@
+     *
+     * @param index An int indicating the position of Task in the ArrayList that is to be marked as done.
+     */
+    public void doneTask(int index) {
```
Should TaskList and Task classes have the same method name here? It might be confusing.

Jeromesyl commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,67 @@
+     * @param command A String containing information that can possibility be used to create an AddCommand object.
+     * @throws duke.DukeException Will be thrown if information provided are insufficient/incorrect.
+     */
+    public AddCommand(String... command) throws duke.DukeException {
```
Perhaps it will be better formatting to use 
```
if (condition) {
    statements;
} else if (condition) {
    statements;
} else {
    statements;
}
``` 
instead of nested
```
if (condition) {
    statements;
} else {
    statements;
}
```

Jeromesyl reviewed (commented):
Very neat and nice formatting overall! I liked how evenly spaced everything is. All public methods had javadocs and most method names are logical. Hardly any coding standard violations.

Jeromesyl commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,116 @@
+            fw.close();
+            Files.delete(filePath);
+            Files.copy(temp, filePath);
+            //noinspection ResultOfMethodCallIgnored
```
What does this comment mean? Should it be included?

jackgugz replied:
You made me wonder also, I think it is a garbage comment I forgot to delete! Thanks HAHA!

Jeromesyl commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,140 @@
+     *
+     * @param index An int indicating the position of Task in the ArrayList that is to be marked as done.
+     */
+    public void doneTask(int index) {
+        Task holder = taskList.get(index);
+        holder.doneTask();
+    }
```
Should TaskList and Task object have the same method name here? It might be confusing.

jackgugz replied:
Hmm a good point from you Jerome so any name you would like to suggest please? Cause it seems regardless what alternative I can currently think of, it will sound pretty similar and still cause the confusion.

Jeromesyl commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,67 @@
+     * @param command A String containing information that can possibility be used to create an AddCommand object.
+     * @throws duke.DukeException Will be thrown if information provided are insufficient/incorrect.
+     */
+    public AddCommand(String... command) throws duke.DukeException {
+        if (command.length == 1) {
+            throw new duke.DukeException(" ☹ OOPS!!! The description of a task cannot be empty.");
+        } else {
+            if (command[0].equals("todo")) {
+                task = new duke.task.Todo(command);
+            } else {
+                if (Arrays.stream(command).anyMatch("/"::equals) && command.length == 2) {
+                    throw new duke.DukeException(
+                            " ☹ HEY DEAR! Please enter a date after / following the task description");
+                }
+                if (command[0].equals("deadline")) {
+                    task = new duke.task.Deadline(command);
+                } else {
+                    task = new duke.task.Event(command);
+                }
+            }
+        }
+    }
```
Perhaps it will be better formatting to use 
```
if (condition) {
    statements;
} else if (condition) {
    statements;
} else {
    statements;
}
``` 
instead of nested
```
if (condition) {
    statements;
} else {
    statements;
}
```

jackgugz replied:
You have a valid point here! Thanks!

jackgugz committed:
Update changes according to commment by Jerome

jackgugz committed:
Fix bug for multiple-word task content
The bug was caused by the introduction of varags and thus the method signature and structures.

jackgugz committed:
Fix bug for JavaFX

Parse method misses parameter and causes the program to fail.

jackgugz committed:
Use assert feature

jackgugz committed:
Merge pull request #1 from jackgugz/branch-A-Assertions

Use assert feature

jackgugz committed:
Improve code quality

jackgugz committed:
Merge pull request #2 from jackgugz/branch-A-CodeQuality

Improve code quality

jackgugz committed:
Merge branch 'master' of https://github.com/jackgugz/ip

jackgugz committed:
Use the Streams feature

jackgugz committed:
Merge branch 'master' into branch-A-Streams

# Conflicts:
#	src/main/java/duke/Storage.java

jackgugz committed:
Set up CI

jackgugz committed:
Merge branch 'master' of https://github.com/jackgugz/ip

jackgugz committed:
Add C-Undo function

jackgugz committed:
Remove last 3 tasks from gradle.yml

jackgugz committed:
Improve the GUI

Let's,
* tweak the GUI to match the asymmetric nature of the conversation
* highlight errors with red text
* tweak padding, fonts, colors, alignments to make the GUI more pleasing to look at
* add profile pictures

jackgugz committed:
Finalize the features

Fix bugs in undo operation for done command
Refactor all files for change of name

jackgugz committed:
Add a representative screenshot of the product to the docs folder

jackgugz committed:
Update README

jackgugz committed:
Set theme jekyll-theme-minimal

jackgugz committed:
Update readme

jackgugz committed:
Update

jackgugz committed:
Update readme link to the jar file

jackgugz committed:
Update on IP:
* Fix bug for GUI misbehaviours on selected PCs
* Fix bug for no error message when entered "delete 1" command when task list is empty
* Update the README and increase readability

jackgugz committed:
Update UG

jackgugz committed:
Fix README

jackgugz committed:
Update README download link

jackgugz committed:
README UPDATE

jackgugz committed:
Fix type in README
